lib/puppet/provider/ntp_config/cisco_ios.rb

```diff
@@ -70,7 +70,24 @@ def update(context, _name, is, should)
       end
     end
 
-    def canonicalize(_context, resources)
+    def canonicalize(context, resources)
+      resources.each do |resource|
+        current_instances = (get(context) || []).find { |key| key[:name] == resource[:name] }
+        # if using unset then we only care if there is currently
+        # source_interface set, if there are none then we don't
+        # need to `unset` them as they do not exist, so removing
+        # the key from the resource
+        if resource[:source_interface] == 'unset'
+          resource.delete(:source_interface) unless current_instances[:source_interface]
+        end
+        # if using [] then we only care if there is currently
+        # trusted_key set, if there are none then we don't
+        # need to `unset` them as they do not exist, so removing
+        # the key from the resource
+        if resource[:trusted_key] == []
+          resource.delete(:trusted_key) unless current_instances[:trusted_key]
+        end
+      end
```

Comment:
same as above - this smacks like `get` is returning the wrong values and this should be sorted there rather than trampling over user-input.

Compare with https://github.com/cisco/cisco-network-puppet-module/blob/1eb6ee9cda566a6f7115f8fdbac72fe264bd923c/lib/puppet/provider/ntp_config/cisco_nexus.rb#L43-L48 for expected behaviours.